flink-ml-lib/src/main/java/org/apache/flink/ml/common/datastream/AllReduceImpl.java

```diff
@@ -0,0 +1,287 @@
+                throw new RuntimeException("The input cannot contain more than one double array.");
+            }
+            hasReceivedOneRecord = true;
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+
+            for (int taskId = 0; taskId < numTasks; taskId++) {
+                int startChunkId = getStartChunkId(taskId, numTasks, receivedElement.length);
+                int numChunksToHandle =
+                        getNumChunksByTaskId(taskId, numTasks, receivedElement.length);
+                for (int chunkId = startChunkId;
+                        chunkId < numChunksToHandle + startChunkId;
+                        chunkId++) {
+                    int bufStart = chunkId * TRANSFER_BUFFER_SIZE;
+                    System.arraycopy(
+                            receivedElement,
+                            bufStart,
+                            transBuf,
+                            0,
+                            getLengthOfChunk(chunkId, receivedElement.length));
+                    out.collect(Tuple3.of(chunkId, receivedElement.length, transBuf));
+                }
+            }
+        }
+    }
+
+    /** Aggregates partitioned array from other workers. */
+    private static class AllReduceSum
+            extends AbstractStreamOperator<Tuple4<Integer, Integer, Integer, double[]>>
+            implements OneInputStreamOperator<
+                            Tuple3<Integer, Integer, double[]>,
+                            Tuple4<Integer, Integer, Integer, double[]>>,
+                    BoundedOneInput {
+
+        /**
+         * A map that aggregates the received partial arrays. The key is chunkId, the value is
+         * (originalArrayLength, aggPartitionedArray).
+         */
+        private Map<Integer, Tuple2<Integer, double[]>> arraySumByChunkId = new HashMap<>();
+
+        @Override
+        public void endInput() {
+            int numTasks = getRuntimeContext().getNumberOfParallelSubtasks();
+            for (Map.Entry<Integer, Tuple2<Integer, double[]>> entry :
+                    arraySumByChunkId.entrySet()) {
+                for (int taskId = 0; taskId < numTasks; taskId++) {
+                    int chunkId = entry.getKey();
+                    int originalArrayLength = entry.getValue().f0;
+                    double[] aggregatedSum = entry.getValue().f1;
+                    output.collect(
+                            new StreamRecord<>(
+                                    Tuple4.of(
+                                            taskId, chunkId, originalArrayLength, aggregatedSum)));
+                }
+            }
+        }
+
+        @Override
+        public void processElement(StreamRecord<Tuple3<Integer, Integer, double[]>> streamRecord) {
+            Tuple3<Integer, Integer, double[]> record = streamRecord.getValue();
+            int chunkId = record.f0;
+            int originalArrayLength = record.f1;
+            double[] partialSum = record.f2;
```

Comment:
The comment at line 70 refers to this field as `partitionedArray`. Should we name the variable `partitionedArray` here for consistency?